system/lib/wasmfs/syscalls.cpp

```diff
@@ -1183,4 +1183,40 @@ int __syscall_poll(intptr_t fds_, int nfds, int timeout) {
   return nonzero;
 }
 
+int __syscall_fallocate(int fd, int mode, uint64_t off, uint64_t len) {
+  assert(mode == 0); // TODO, but other modes were never supported in the old FS
+
+  auto fileTable = wasmFS.getFileTable().locked();
+  auto openFile = fileTable.getEntry(fd);
+  if (!openFile) {
+    return -EBADF;
+  }
+
+  auto dataFile = openFile->locked().getFile()->dynCast<DataFile>();
+  // TODO: support for symlinks.
+  if (!dataFile) {
+    return -ENODEV;
+  }
+
+  auto locked = dataFile->locked();
+  if (!(locked.getMode() & WASMFS_PERM_WRITE)) {
+    return -EBADF;
+  }
+
+  if (off < 0 || len <= 0) {
+    return -EINVAL;
+  }
+
+  // TODO: we silently do nothing if the stream does not support allocation, but
+  //       in principle we should return EOPNOTSUPP
+  // TODO: instead of just setting the size, we could only increase the size if
+  //       off + len is greater than the old size, and we could only fill zeros
+  //       for regions that were completely unused before (for a backend with
+  //       sparse data storage, that would make a difference from the current
+  //       code).
+  locked.setSize(off + len);
```

Comment:
Shouldn't we check to make sure we only increase size? I know it's mentioned in the TODO, but this seems like a bad correctness problem, not just missing functionality.

Reply:
Ah, good catch. I initially added a new backend API, then removed it when I realized it was not needed, but I did not refactor this properly after that...